File: src/runtime/restore.ts
import * as path from 'path';
import * as fs from 'fs';
import { fork } from 'child_process';
import { readSession } from '../pty/pty-manager';
import { tailBuffer } from '../buffer/buffer-manager';
import { logEvent } from '../ledger/ledger';

/**
 * Restore a suspended session by spawning a new worker with --restore.
 * Waits for the worker to initialize and returns when ready.
 */
export async function restoreSession(
  terminalId: string,
  projectRoot: string
): Promise<void> {
  const session = readSession(terminalId, projectRoot);
  if (!session || session.status !== 'suspended') {
    throw new Error(`Session ${terminalId} is not suspended`);
  }

  // Prefer saved cwd, fall back to the session's original cwd.
  const restoredCwd = session.saved_state?.cwd ?? session.cwd;

  // Find the worker script
  const workerScript = path.join(__dirname, '..', 'worker.js');
  const workerScriptTs = path.join(__dirname, '..', 'worker.ts');
  const script = fs.existsSync(workerScript) ? workerScript : workerScriptTs;

  const child = fork(
    script,
    [terminalId, session.command, restoredCwd, projectRoot, '--restore'],
    {
      detached: true,
      stdio: 'ignore',
      execArgv: script.endsWith('.ts') ? ['--import', 'tsx'] : [],
    }
  );

  child.unref();

  logEvent('session.restored', projectRoot, terminalId, {
    worker_pid: child.pid,
    restored_cwd: restoredCwd,
  });

  // Wait for the worker to come up and the session to become running
  const maxWait = 3000;
  const poll = 100;
  let elapsed = 0;

  while (elapsed < maxWait) {
    await new Promise((r) => setTimeout(r, poll));
    elapsed += poll;

    const updated = readSession(terminalId, projectRoot);
    if (updated && updated.status === 'running' && updated.worker_pid !== session.worker_pid) {
      return; // Worker is up
    }
  }
}

File: python/clrun/runtime/restore.py
"""Restore a suspended session by spawning a new worker with --restore."""

from __future__ import annotations

import os
import subprocess
import sys
import time

from clrun.pty.pty_manager import read_session
from clrun.ledger.ledger import log_event


def restore_session(terminal_id: str, project_root: str) -> None:
    session = read_session(terminal_id, project_root)
    if not session or session.status != "suspended":
        raise RuntimeError(f"Session {terminal_id} is not suspended")

    restored_cwd = session.saved_state.cwd if session.saved_state else session.cwd

    worker_module = "clrun.worker"
    child = subprocess.Popen(
        [
            sys.executable, "-m", worker_module,
            terminal_id, session.command, restored_cwd, project_root, "--restore",
        ],
        start_new_session=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        stdin=subprocess.DEVNULL,
    )

    log_event("session.restored", project_root, terminal_id, {
        "worker_pid": child.pid,
        "restored_cwd": restored_cwd,
    })

    max_wait = 3.0
    poll_interval = 0.1
    elapsed = 0.0

    while elapsed < max_wait:
        time.sleep(poll_interval)
        elapsed += poll_interval
        updated = read_session(terminal_id, project_root)
        if updated and updated.status == "running" and updated.worker_pid != session.worker_pid:
            return
